File: LC_PYTHON/Kth_max_and_min_array.py
#User function Template for python3

import heapq

class Solution:

    def kthSmallest(self, arr,k):
        arr1 = []
        for i in arr:
            if len(arr1) == k:
                if heapq.nlargest(1, arr1)[0] > i:
                    heapq.heappop(arr1)                    
                    heapq.heappush(arr1, i)
                    heapq._heapify_max(arr1)
            else:
                heapq.heappush(arr1, i)
                heapq._heapify_max(arr1)
        return heapq.nlargest(1, arr1)[0]
        

#{ 
 # Driver Code Starts
#Initial Template for Python 3

#contributed by RavinderSinghPB
if __name__ == '__main__':
    import random
    t = int(input())
    for tcs in range(t):
        # n = int(input())
        arr = list(map(int, input().strip().split()))
        k = int(input())
        ob = Solution()
        print(ob.kthSmallest(arr, k))

# } Driver Code Ends

File: LC_JAVA/Kth_max_and_min_array.java
//{ Driver Code Starts
// Initial Template for Java

/*package whatever //do not write package name here */

import java.io.*;
import java.util.*;
import java.util.PriorityQueue;

class Kth_max_and_min_array {
    public static void main(String args[]) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(System.out);

        int t = Integer.parseInt(in.readLine().trim());
        while (t-- > 0) {
            String line = in.readLine();
            String[] tokens = line.split(" ");

            // Create an ArrayList to store the integers
            ArrayList<Integer> array = new ArrayList<>();

            // Parse the tokens into integers and add to the array
            for (String token : tokens) {
                array.add(Integer.parseInt(token));
            }

            int[] arr = new int[array.size()];
            int idx = 0;
            for (int i : array) arr[idx++] = i;

            int key = Integer.parseInt(in.readLine().trim());
            Solution ob = new Solution();
            out.println(ob.kthSmallest(arr, key));
        }
        out.flush();
    }
}

// } Driver Code Ends


// User function Template for Java

class Solution {
    public static int kthSmallest(int[] arr, int k) {
        // Your code here

        PriorityQueue pq= new PriorityQueue<Integer>(Comparator.reverseOrder()); //max heap
        for(int i:arr){
            if(pq.size()==k ){
                if(Integer.parseInt(pq.peek().toString()) > i){
                    pq.poll();
                    pq.add(i);   
                }                         
            }
            else{
                pq.add(i);
            }
        }

        return Integer.parseInt(pq.peek().toString());
    }
}
